docs/concurrency.md

```diff
@@ -89,24 +91,26 @@ To formalize the difference between general and acyclic dataflow, we note that a
-The three streams are defined recursively: the `View` is a function of the `Model`; the `Stream[Model]` is a function of `Stream[Input]` (user input events update the model); and `Input` also depends on `View` since the `View` determines what control elements are shown on the screen at the time, and thus what input events the user can create.
+The types `Model`, `View`, and `Input` stand for data that represent the type of the data model of the application; the type of the entire view (all windows) shown on the screen; and all the possible input events that the user might create (including keyboard and mouse).
+The three streams are defined recursively: the `View` is a function of the `Model`; the `Stream[Model]` is a function of `Stream[Input]` (some input events will update the model); and `Input` also depends on `View` since the `View` determines which control elements are shown on the screen at any time, and thus determines what input events the user can create at that time.
 
 The streaming frameworks that do not support a recursive definition of streams fall into the acyclic dataflow class.
 
 ## Level 4: General concurrency
 
 Finally, we consider the general concurrency problems.
 The main task in this class of problems is to manage several computations that run concurrently in unknown order and interact in arbitrary ways.
-For instance, one computation may at some point stop and wait until another computation computes a certain result, and then examine that result to decide whether to continue its own thread of computation or wait further.
+For instance, one computation may at some point stop and wait until another computation computes a certain result, and then examine that result to decide whether to continue its own thread of computation, or wait further, or create new computation threads.
 
 The main difficulty here is to ensure that different processes are synchronized in the desired manner.
 
 Frameworks such as Akka Actors, Go coroutines/channels, and the Java concurrency primitives (`Thread`, `wait/notify`, `synchronized`) are all Level 4 concurrency frameworks.
 The chemical machine (known in the academic world as "join calculus") is also a Level 4 framework.
 
-The typical task that requires this level of concurrency is implementing an operating system.
-The "dining philosophers" problem is also an example of a concurrency task that cannot be implemented by any concurrency framework other than a Level 4 framework.
+The typical task that requires this level of concurrency is implementing an operating system where many running processes can synchronize and communicate with each other in arbitrary ways.
+
+It seems to me that the "dining philosophers" problem is also an example of a concurrency task that cannot be implemented by any concurrency framework other than a Level 4 framework.
+However, I do not know how to prove that this is so.
 
 ### Why is Level 4 higher than Level 3
```

Comment:
If you can slice processes A and B into some smaller parts and then interleave these parts with a scheduler on a single thread, you have essentially implemented concurrency with 2 threads. For this, both A and B must have some interfaces that will allow them to be sliced up.  I assumed that you can only run A and B, but not slice them up or modify them in any way. For instance, Java functions cannot be sliced up using only Java APIs. Continuations are not going to help you because, to use continuations, the processes A and B need to be rewritten so that they call a certain continuation just before starting a potentially infinite loop (to give the scheduler a chance to schedule the other process), or something like that.

Of course, MS DOS could do this on an x86 using interrupts, since a scheduled interrupt will stop anything that runs on the CPU.  I seem to remember some early MS DOS enhancement that offered true preemptive multitasking in this way. I am not sure what is the formal description of interrupts, but if you allow them, there is no distinction between one thread and many threads any more. This distinction becomes the distinction between having an interrupt-based scheduler or not having it.